test/test_foreach.py

```diff
@@ -119,6 +120,96 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
                    torch.tensor([1], dtype=torch.long, device=device)]
         self.assertRaises(RuntimeError, lambda: torch._foreach_add(tensors, 1))
 
+    def test_add_list_error_cases(self, device):
+        tensors1 = []
+        tensors2 = []
+
+        # Empty lists
+        with self.assertRaises(RuntimeError):
+            torch._foreach_add(tensors1, tensors2)
+        with self.assertRaises(RuntimeError):
+            torch._foreach_add_(tensors1, tensors2)
+
+        # One empty list
+        tensors1.append(torch.tensor([1], device=device))
+        with self.assertRaisesRegex(RuntimeError, "Tensor list must have at least one tensor."):
+            torch._foreach_add(tensors1, tensors2)
+        with self.assertRaisesRegex(RuntimeError, "Tensor list must have at least one tensor."):
+            torch._foreach_add_(tensors1, tensors2)
+
+        # Lists have different amount of tensors
+        tensors2.append(torch.tensor([1], device=device))
+        tensors2.append(torch.tensor([1], device=device))
+        with self.assertRaisesRegex(RuntimeError, "Tensor lists must have the same number of tensors, got 1 and 2"):
+            torch._foreach_add(tensors1, tensors2)
+        with self.assertRaisesRegex(RuntimeError, "Tensor lists must have the same number of tensors, got 1 and 2"):
+            torch._foreach_add_(tensors1, tensors2)
+
+        # Different dtypes
+        tensors1 = [torch.zeros(10, 10, device=device, dtype=torch.float) for _ in range(10)]
+        tensors2 = [torch.ones(10, 10, device=device, dtype=torch.int) for _ in range(10)]
+
+        with self.assertRaisesRegex(RuntimeError, "All tensors in the tensor list must have the same dtype."):
+            torch._foreach_add(tensors1, tensors2)
+        with self.assertRaisesRegex(RuntimeError, "All tensors in the tensor list must have the same dtype."):
+            torch._foreach_add_(tensors1, tensors2)
+
+        # different devices
+        tensor1 = torch.zeros(10, 10, device="cuda:0")
+        tensor2 = torch.ones(10, 10, device="cuda:1")
+        with self.assertRaisesRegex(RuntimeError, "Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!"):
```

Comment:
Feel free to just check part of the string, e.g. "Expected all tensors to be on the same device". The goal of checking assertRaisesRegex is to make sure the error message isn't something completely unreadable like "TORCH_INTERNAL_ASSERT(false)".